src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,259 @@
+        {
+            if (uri->empty())
+            {
+                sleDID->makeFieldAbsent(sfURI);
+            }
+            else
+            {
+                (*sleDID)[sfURI] = *uri;
+            }
+        }
+        if (auto const document = ctx_.tx[~sfDIDDocument])
+        {
+            if (document->empty())
+            {
+                sleDID->makeFieldAbsent(sfDIDDocument);
+            }
+            else
+            {
+                (*sleDID)[sfDIDDocument] = *document;
+            }
+        }
+
+        if (auto const attestation = ctx_.tx[~sfAttestation])
+        {
+            if (attestation->empty())
+            {
+                sleDID->makeFieldAbsent(sfAttestation);
+            }
+            else
+            {
+                (*sleDID)[sfAttestation] = *attestation;
+            }
+        }
+        if (!sleDID->isFieldPresent(sfURI) &&
+            !sleDID->isFieldPresent(sfDIDDocument))
+        {
+            return tecEMPTY_DID;
+        }
+        ctx_.view().update(sleDID);
+        return tesSUCCESS;
+    }
+
+    // Create new ledger object otherwise
+    auto const sleDID = std::make_shared<SLE>(didKeylet);
+    (*sleDID)[sfAccount] = account_;
+    if (auto const uri = ctx_.tx[~sfURI]; uri.has_value() && !uri->empty())
+        (*sleDID)[sfURI] = uri.value();
+    if (auto const document = ctx_.tx[~sfDIDDocument];
+        document.has_value() && !document->empty())
+        (*sleDID)[sfDIDDocument] = document.value();
+    if (auto const attestation = ctx_.tx[~sfAttestation];
+        attestation.has_value() && !attestation->empty())
+        (*sleDID)[sfAttestation] = attestation.value();
+
+    return addSLE(ctx_, sleDID, account_);
+}
+
+NotTEC
+DIDDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureDID))
+        return temDISABLED;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    return preflight2(ctx);
+}
+
+TER
+DIDDelete::deleteSLE(ApplyContext& ctx, Keylet sleKeylet, AccountID const owner)
+{
+    auto const sle = ctx.view().peek(sleKeylet);
+    if (!sle)
+        return tecNO_ENTRY;
+
+    return DIDDelete::deleteSLE(ctx.view(), sle, owner, ctx.journal);
+}
+
+TER
+DIDDelete::deleteSLE(
+    ApplyView& view,
+    std::shared_ptr<SLE> sle,
+    AccountID const owner,
+    beast::Journal j)
+{
+    // Remove object from owner directory
+    if (!view.dirRemove(
+            keylet::ownerDir(owner), (*sle)[sfOwnerNode], sle->key(), true))
+    {
+        JLOG(j.fatal()) << "Unable to delete DID Token from owner.";
+        return tefBAD_LEDGER;
+    }
+
+    auto const sleOwner = view.peek(keylet::account(owner));
+    adjustOwnerCount(view, sleOwner, -1, j);
+    view.update(sleOwner);
+
+    // Remove object from ledger
+    view.erase(sle);
+    return tesSUCCESS;
+}
+
+TER
+DIDDelete::doApply()
+{
```

Comment:
`account_` is equal to `ctx_.tx[sfAccount]`. Consider this:
>     return deleteSLE(ctx_, keylet::did(account_), account_);

Reply:
Done in https://github.com/XRPLF/rippled/pull/4636/commits/298077f443ad9e0a0a8ee3045f2dafd0bd2a590b